src/app-layer-dhcp.c

```diff
@@ -0,0 +1,668 @@
+ *
+ * \retval ALPROTO_DHCP if it looks like echo, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto dhcpToServerProbingParser(uint8_t *input, uint32_t input_len,
+    uint32_t *offset)
+{
+    /* TBD: have the infrastructure call us back with the flow struct *
+     * so that we can check that this arrived on the proper 5 tuple
+     */
+    //PrintRawDataFp(stdout, input, input_len);
+
+    if (input_len >= DHCP_MIN_FRAME_LEN) {
+        BOOTPHdr *bootp = (BOOTPHdr *)input;
+
+        if ((bootp->op == BOOTP_REQUEST) &&
+            (bootp->htype == BOOTP_ETHERNET) &&
+            (bootp->hlen == 6) &&
+            (bootp->magic == ntohl(BOOTP_DHCP_MAGIC_COOKIE))) {
+            DHCPOpt *dhcp = (DHCPOpt *)(input + sizeof(BOOTPHdr));
+
+            if ((dhcp->code == DHCP_DHCP_MSG_TYPE) &&
+                (dhcp->len == 1)) {
+
+                SCLogDebug("Detected as ALPROTO_DHCP.");
+                return ALPROTO_DHCP;
+            }
+        }
+    }
+
+    SCLogDebug("Protocol not detected as ALPROTO_DHCP.");
+    return ALPROTO_UNKNOWN;
+}
+
+/**
+ * \brief Probe the input to see if it looks like echo.
+ *
+ * \retval ALPROTO_DHCP if it looks like echo, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto dhcpToClientProbingParser(uint8_t *input, uint32_t input_len,
+    uint32_t *offset)
+{
+    /* TBD: have the infrastructure call us back with the flow struct *
+     * so that we can check that this arrived on the proper 5 tuple
+     */
+    //PrintRawDataFp(stdout, input, input_len);
+
+    if (input_len >= DHCP_MIN_FRAME_LEN) {
+        BOOTPHdr *bootp = (BOOTPHdr *)input;
+
+        if ((bootp->op == BOOTP_REPLY) &&
+            (bootp->htype == BOOTP_ETHERNET) &&
+            (bootp->hlen == 6) &&
+            (bootp->magic == ntohl(BOOTP_DHCP_MAGIC_COOKIE))) {
+            DHCPOpt *dhcp = (DHCPOpt *)(input + sizeof(BOOTPHdr));
+
+            if ((dhcp->code == DHCP_DHCP_MSG_TYPE) &&
+                (dhcp->len == 1)) {
+
+                SCLogDebug("Detected as ALPROTO_DHCP.");
+                return ALPROTO_DHCP;
+            }
+        }
+    }
+
+    SCLogDebug("Protocol not detected as ALPROTO_DHCP.");
+    return ALPROTO_UNKNOWN;
+}
+
+static int dhcpParse(Flow *f, void *state,
+    AppLayerParserState *pstate, uint8_t *input, uint32_t input_len,
+    void *local_data)
+{
+    dhcpState *dhcp_state = state;
+
+    /* Likely connection closed, we can just return here. */
+    if ((input == NULL || input_len == 0) &&
+        AppLayerParserStateIssetFlag(pstate, APP_LAYER_PARSER_EOF)) {
+        return 0;
+    }
+
+    /* Probably don't want to create a transaction in this case
+     * either. */
+    if (input == NULL || input_len == 0) {
+        return 0;
+    }
+
+    /* Normally you would parse out data here and store it in the
+     * transaction object, but as this is echo, we'll just record the
+     * request data. */
+
+    if (input_len >= DHCP_MIN_FRAME_LEN) {
```

Comment:
if (input_len < DHCP_MIN_FRAME_LEN)\n goto end;
would have the same effect, except the whole block after can move one step left. Much preferred for readability.